coordinator/internal/transitengine/transitengine.go

```diff
@@ -0,0 +1,244 @@
+			return
+		}
+		decReq, err := parseDecryptionRequest(r)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("parsing decryption request: %v", err), http.StatusBadRequest)
+			return
+		}
+
+		key, err := deriveEncryptionKey(authority, fmt.Sprintf("%d_", decReq.CiphertextContainer.keyVersion)+workloadSecretID)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("key derivation: %v", err), http.StatusInternalServerError)
+			return
+		}
+		plaintext, err := symmetricDecryptRaw(key, decReq.CiphertextContainer, decReq.AssociatedData)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("decrypting: %v", err), http.StatusInternalServerError)
+			return
+		}
+		var decResp decryptionResponse
+		decResp.Plaintext = plaintext
+		if err = writeJSONResponse(w, decResp); err != nil {
+			http.Error(w, fmt.Sprintf("writing response: %v", err), http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+// deriveEncryptionKey derives the workload secret used as the encryption key by receiving the seedengine of the current state.
+func deriveEncryptionKey(authority stateAuthority, workloadSecretID string) ([]byte, error) {
+	state, err := authority.GetState()
+	if err != nil {
+		return nil, err
+	}
+	// TODO(jmxnzo): authentication of client certs <-> parsed workloadSecretID.
+	derivedWorkloadSecret, err := state.SeedEngine.DeriveWorkloadSecret(workloadSecretID)
+	if err != nil {
+		return nil, err
+	}
+	if len(derivedWorkloadSecret) < aesGCMKeySize {
+		return nil, fmt.Errorf("derived key too small, expected key length: %d", aesGCMKeySize)
+	}
+	return derivedWorkloadSecret[:aesGCMKeySize], nil
+}
+
+// writeJSONResponse wraps any payload inside a "data" object and sends it as an HTTP response.
+func writeJSONResponse(w http.ResponseWriter, payload any) error {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	response := map[string]any{
+		"data": payload,
+	}
+	return json.NewEncoder(w).Encode(response)
+}
+
+// parseEncryptionRequest returns the given HTTP request body as encryptionRequest holding
+// the b64Plaintext and the optional query parameter associatedData.
+func parseEncryptionRequest(r *http.Request) (encryptionRequest, error) {
+	defer r.Body.Close()
+	if err := validateContentType(r); err != nil {
+		return encryptionRequest{}, err
+	}
+	var encReq encryptionRequest
+	if err := json.NewDecoder(r.Body).Decode(&encReq); err != nil {
+		return encryptionRequest{}, err
+	}
+	return encReq, nil
+}
+
+// parseDecryptionRequest returns the given HTTP request body as decryptionRequest holding
+// the ciphertextContainer and the optional query parameter associatedData.
+func parseDecryptionRequest(r *http.Request) (decryptionRequest, error) {
+	defer r.Body.Close()
+	if err := validateContentType(r); err != nil {
+		return decryptionRequest{}, err
+	}
+	var decReq decryptionRequest
+	if err := json.NewDecoder(r.Body).Decode(&decReq); err != nil {
+		return decryptionRequest{}, err
+	}
+	return decReq, nil
+}
```

Comment:
I think this could be simplified like this:

```go
func parseRequest(r *http.Request, into *any) error {
  // ...
}
// At the call site:
var decReq decryptionRequest
parseRequest(r, &decReq)
```